### PR TITLE
feat: Add support for bgp dynamic neighbours subcommand for metal gateway

### LINKED ISSUE #413: Add support for Metal Gateway BGP Dynamic Neighbors
### What problem are you facing?

For complete Metal Gateway support (#205) there are a few new features of Metal Gateway yet to be implemented. Among them is BGP Dynamic Neighbors. This also ties to VRF (#268) as the features are closely associated.

https://github.com/equinix-labs/metal-go/blob/main/docs/VRFsApi.md

Method | HTTP request | Description
------------- | ------------- | -------------
[**BgpDynamicNeighborsIdGet**](https://github.com/equinix-labs/metal-go/blob/main/docs/VRFsApi.md#BgpDynamicNeighborsIdGet) | **Get** /bgp-dynamic-neighbors/{id} | Retrieve a BGP Dynamic Neighbor
[**CreateBgpDynamicNeighbor**](https://github.com/equinix-labs/metal-go/blob/main/docs/VRFsApi.md#CreateBgpDynamicNeighbor) | **Post** /metal-gateways/{id}/bgp-dynamic-neighbors | Create a VRF BGP Dynamic Neighbor range
[**DeleteBgpDynamicNeighborById**](VRFsApi.md#DeleteBgpDynamicNeighborById) | **Delete** /bgp-dynamic-neighbors/{id} | Delete a VRF BGP Dynamic Neighbor
[**GetBgpDynamicNeighbors**](https://github.com/equinix-labs/metal-go/blob/main/docs/VRFsApi.md#GetBgpDynamicNeighbors) | **Get** /metal-gateways/{id}/bgp-dynamic-neighbors | List BGP Dynamic Neighbors

### How could the Equinix Metal CLI help solve this problem?

```
metal gateway list-bgp-dynamic-neighbors --id <gateway_id>
metal gateway create-bgp-dynamic-neighbors --id <gateway_id>
metal gateway delete-bgp-dynamic-neighbors --bgp_dynamic_neighbor_id <bgp_dynamic_neighbor_id>
metal gateway get-bgp-dynamic-neighbors --bgp_dynamic_neighbor_id <bgp_dynamic_neighbor_id>
```

These command names are wordy, open to suggestions.

### PR DESCRIPTION
**Pull request by aayushrangwala**
Fixes: https://github.com/equinix/metal-cli/issues/413

### CODE REVIEW

File: internal/gateway/bgp-dynamic-neighbours/create.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/spf13/cobra"
+)
+
+func (c *Client) Create() *cobra.Command {
+	var gatewayId, bgpNeighbourRange string
+	var asn int32
+
+	// createGwBgpCmd represents the creation of gateway bgp dynamic neighbour command
+	createGwBgpCmd := &cobra.Command{
+		Use:     ``,
```
Comment (displague) [inline internal/gateway/bgp-dynamic-neighbours/create.go]: We'll need these doc fields filled in. Converting to draft until ready.

Comment (aayushrangwala) [reply internal/gateway/bgp-dynamic-neighbours/create.go]: Yes, not complete yet. Will be doing those changes

**Review by ctreatma — CHANGES_REQUESTED** [review]
Can you add tests for this subcommand?

**Comment by aayushrangwala** [comment]
> Can you add tests for this subcommand?

@ctreatma I am working on the tests but should we keep small PRs and raise tests in the next one? It'll be easier to review as well. And anyway this is a new functionality which is not release

**Comment by ctreatma** [comment]
> @ctreatma I am working on the tests but should we keep small PRs and raise tests in the next one? It'll be easier to review as well. And anyway this is a new functionality which is not release

Tests build the reviewer's confidence that the changes actually work the way they are supposed to.  Doing them in a separate PR only serves to delay finishing the code.

File: internal/gateway/getbgpneighbours.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	"github.com/spf13/cobra"
+)
+
+func (c *Client) GetBgpNeighbours() *cobra.Command {
+	var bgpNeighbourID string
+
+	// getGwayBgpDynamicNeighbourCmd represents command to get Metal Gateway Dynamic Neighbour by ID.
+	getGwayBgpDynamicNeighbourCmd := &cobra.Command{
+		Use:   `get-bgp-dynamic-neighbours`,
```
Comment (ctreatma) [inline internal/gateway/getbgpneighbours.go]: Equinix docs/tools/etc. generally use American English spellings, so we should align with that throughout this PR.

```suggestion
		Use:   `get-bgp-dynamic-neighbors`,
```

File: test/e2e/gateways/bgp-dynamic-neighbour/create_test.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	metal "github.com/equinix/equinix-sdk-go/services/metalv1"
+	"github.com/spf13/cobra"
+
+	root "github.com/equinix/metal-cli/internal/cli"
+	"github.com/equinix/metal-cli/internal/gateway"
+	outputPkg "github.com/equinix/metal-cli/internal/outputs"
+	"github.com/equinix/metal-cli/test/helper"
+)
+
+func TestBgpDynamicNeighbours_Create(t *testing.T) {
```
Comment (ctreatma) [inline test/e2e/gateways/bgp-dynamic-neighbour/create_test.go]: All of the new tests [are failing with 404s](https://github.com/equinix/metal-cli/actions/runs/7672035602/job/20912047460#step:5:114):

```
=== RUN   TestBgpDynamicNeighbours_Create/create_bgp_dynamic_neighbour
Error: Could not create BGP Dynamic Neighbour: 404 Not Found Not found
    helper.go:591: Could not create BGP Dynamic Neighbour: 404 Not Found Not found
--- FAIL: TestBgpDynamicNeighbours_Create (2.54s)
    --- FAIL: TestBgpDynamicNeighbours_Create/create_bgp_dynamic_neighbour (0.14s)
=== RUN   TestBgpDynamicNeighbours_Delete
    delete_test.go:27: Error when calling `VRFsApi.CreateBgpDynamicNeighbor`: 404 Not Found Not found
--- FAIL: TestBgpDynamicNeighbours_Delete (3.10s)
=== RUN   TestBgpDynamicNeighbours_Get
    get_test.go:25: Error when calling `VRFsApi.CreateBgpDynamicNeighbor`: 404 Not Found Not found
--- FAIL: TestBgpDynamicNeighbours_Get (2.65s)
=== RUN   TestBgpDynamicNeighbours_List
    list_test.go:25: Error when calling `VRFsApi.CreateBgpDynamicNeighbor`: 404 Not Found Not found
--- FAIL: TestBgpDynamicNeighbours_List (3.00s)
FAIL
```

Do we need to pass in any additional CLI flags?

File: internal/gateway/listbgpneighbours.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+				data[0] = []string{neighbour.GetId(), neighbour.GetBgpNeighborRange(),
+					strconv.Itoa(int(neighbour.GetBgpNeighborAsn())), string(neighbour.GetState()), neighbour.GetCreatedAt().String()}
+			}
+
+			header := []string{"ID", "Range", "ASN", "State", "Created"}
+			return c.Out.Output(n, header, &data)
+		},
+	}
+
+	createMetalGatewayCmd.Flags().StringVar(&gatewayId, "gateway-id", "", "")
```
Comment (ctreatma) [inline internal/gateway/listbgpneighbours.go]: This should be `id` to match the docs

```suggestion
	createMetalGatewayCmd.Flags().StringVar(&gatewayId, "id", "", "")
```

File: internal/gateway/getbgpneighbours.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		Short: "Gets a BGP Dynamic Neighbour",
+		Long:  "Gets the BGP Dynamic Neighbour for the metal gateway with the specified ID",
```
Comment (ctreatma) [inline internal/gateway/getbgpneighbours.go]: I think `metal gateway` shouldn't be mentioned here?

File: internal/gateway/getbgpneighbours.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+			data[0] = []string{n.GetId(), n.GetBgpNeighborRange(),
+				strconv.Itoa(int(n.GetBgpNeighborAsn())), string(n.GetState()), n.GetCreatedAt().String()}
+			header := []string{"ID", "Range", "ASN", "State", "Created"}
+
+			return c.Out.Output(n, header, &data)
+		},
+	}
+
+	getGwayBgpDynamicNeighbourCmd.Flags().StringVarP(&bgpNeighbourID, "id", "i", "", "BGP Dynamic Neighbour ID. Ex: []")
```
Comment (ctreatma) [inline internal/gateway/getbgpneighbours.go]: I think we talked about using `id` for the gateway ID and `bgp-neighbor-id` for the neighbor ID, since these commands are part of the `gateway` subcommand.

File: test/e2e/gateways/bgp-dynamic-neighbour/create_test.go (hunk 3 above)
Comment (aayushrangwala) [reply test/e2e/gateways/bgp-dynamic-neighbour/create_test.go]: Not sure why its failing with 404 on the CI, its passing at local

```
~/go/src/metal-cli   gw-bgp-neighbours *2 !15 ?4 ❯ go test -v ./test/e2e/gateways/bgp-dynamic-neighbour/...                                                                          27s
=== RUN   TestBgpDynamicNeighbors_Create
=== RUN   TestBgpDynamicNeighbors_Create/create_bgp_dynamic_neighbor
--- PASS: TestBgpDynamicNeighbors_Create (6.15s)
    --- PASS: TestBgpDynamicNeighbors_Create/create_bgp_dynamic_neighbor (0.83s)
=== RUN   TestBgpDynamicNeighbors_Delete
=== RUN   TestBgpDynamicNeighbors_Delete/delete_bgpDynamicNeighbor_by_ID
--- PASS: TestBgpDynamicNeighbors_Delete (6.11s)
    --- PASS: TestBgpDynamicNeighbors_Delete/delete_bgpDynamicNeighbor_by_ID (0.82s)
=== RUN   TestBgpDynamicNeighbors_Get
=== RUN   TestBgpDynamicNeighbors_Get/get_bgpDynamicNeighbor_by_ID
--- PASS: TestBgpDynamicNeighbors_Get (5.97s)
    --- PASS: TestBgpDynamicNeighbors_Get/get_bgpDynamicNeighbor_by_ID (0.43s)
=== RUN   TestBgpDynamicNeighbors_List
=== RUN   TestBgpDynamicNeighbors_List/list_bgpDynamicNeighbor_by_ID
--- PASS: TestBgpDynamicNeighbors_List (6.04s)
    --- PASS: TestBgpDynamicNeighbors_List/list_bgpDynamicNeighbor_by_ID (0.66s)
PASS
ok  	github.com/equinix/metal-cli/test/e2e/gateways/bgp-dynamic-neighbour	24.716s
```

File: internal/gateway/getbgpneighbours.go (hunk 6 above)
Comment (aayushrangwala) [reply internal/gateway/getbgpneighbours.go]: Yes fixed

File: internal/gateway/getbgpneighbours.go (hunk 5 above)
Comment (aayushrangwala) [reply internal/gateway/getbgpneighbours.go]: Removed

File: internal/gateway/listbgpneighbours.go (hunk 4 above)
Comment (aayushrangwala) [reply internal/gateway/listbgpneighbours.go]: fixed

File: internal/gateway/getbgpneighbours.go (hunk 2 above)
Comment (aayushrangwala) [reply internal/gateway/getbgpneighbours.go]: Corrected at all the references

**Comment by displague** [comment]
I enabled VRF over Shared Connections, VRF Static Routers, and BGP Dynamic Neighbors for the testing org. We should see better progress in the E2E.

File: test/helper/helper.go [hunk 7]
```diff
@@ -670,8 +698,8 @@ func CleanTestVrfs(t *testing.T, vrfId string) {
 	t.Helper()
 	TestApiClient := TestClient()
 
-	resp, err := TestApiClient.VRFsApi.DeleteVrf(context.Background(), vrfId).Execute()
-	if err != nil && resp.StatusCode != http.StatusNotFound {
+	_, err := TestApiClient.VRFsApi.DeleteVrf(context.Background(), vrfId).Execute()
+	if err != nil && strings.Contains(err.Error(), "Not Found") {
```
Comment (ctreatma) [inline test/helper/helper.go]: At a glance I would expect both the old and new code to work here, but the status code check is more explicit.  Could you change it back and we'll see if that fixes the test cleanup failure?

```
=== RUN   TestCli_Vrf_Create/vrf-delete-test
    helper.go:690: Error when calling `VRFsApi.DeleteVrf`` for 548a73da-78b0-484f-b7be-5c6d5e48e714: 404 Not Found Not found
```

**Comment by aayushrangwala** [comment]
> strings.Contains(err.Error(), "Not Found")

Yes after enabling FFs, tests started passing for BGP 
https://github.com/equinix/metal-cli/actions/runs/7700595014/job/20986048893?pr=433#step:5:111

File: test/helper/helper.go (hunk 7 above)
Comment (aayushrangwala) [reply test/helper/helper.go]: Yes reverted

File: test/helper/helper.go [hunk 8]
```diff
@@ -642,14 +652,14 @@ func waitForInterconnectionDeleted(apiClient *metalv1.APIClient, connId string,
 }
 
 //nolint:staticcheck
-func CreateTestVrfs(t *testing.T, projectId, name string, vlan int) *metalv1.Vrf {
+func CreateTestVrfs(t *testing.T, projectId, name string, vlan int32) *metalv1.Vrf {
```
Comment (cprivitere) [inline test/helper/helper.go:657]: What is the purpose of this change?

Comment (cprivitere) [reply test/helper/helper.go:657]: I figured it out. Ignore this.

**Comment by cprivitere** [comment]
We're having issues getting the device create tests to pass, but I firmly believe that's unrelated to this PR and we can improve those in a separate PR. Going to merge now.

File: test/e2e/gateways/bgp-dynamic-neighbour/create_test.go (hunk 3 above)
Comment (cprivitere) [reply test/e2e/gateways/bgp-dynamic-neighbour/create_test.go]: I'm also getting the 404 Not found Not Found on my local tests.

**Comment by aayushrangwala** [comment]
```
=== RUN   TestBgpDynamicNeighbors_Create
=== RUN   TestBgpDynamicNeighbors_Create/create_bgp_dynamic_neighbor
--- PASS: TestBgpDynamicNeighbors_Create (5.21s)
    --- PASS: TestBgpDynamicNeighbors_Create/create_bgp_dynamic_neighbor (0.45s)
=== RUN   TestBgpDynamicNeighbors_Delete
=== RUN   TestBgpDynamicNeighbors_Delete/delete_bgpDynamicNeighbor_by_ID
--- PASS: TestBgpDynamicNeighbors_Delete (3.82s)
    --- PASS: TestBgpDynamicNeighbors_Delete/delete_bgpDynamicNeighbor_by_ID (0.36s)
=== RUN   TestBgpDynamicNeighbors_Get
=== RUN   TestBgpDynamicNeighbors_Get/get_bgpDynamicNeighbor_by_ID
--- PASS: TestBgpDynamicNeighbors_Get (3.22s)
    --- PASS: TestBgpDynamicNeighbors_Get/get_bgpDynamicNeighbor_by_ID (0.19s)
=== RUN   TestBgpDynamicNeighbors_List
=== RUN   TestBgpDynamicNeighbors_List/list_bgpDynamicNeighbor_by_ID
--- PASS: TestBgpDynamicNeighbors_List (3.61s)
    --- PASS: TestBgpDynamicNeighbors_List/list_bgpDynamicNeighbor_by_ID (0.19s)
```

All related tests are passing in this test run @ctreatma can you please approve this PR

**Review by ctreatma — APPROVED** [review]
The only test failure in CI was a panic due to hitting the go test timeout.  I'm going to assume it's unrelated and deal with any consequences later.
